### PR TITLE
GitHub Actions: cache composer dependencies

### PR DESCRIPTION
**Pull request by bendavies**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

GitHub Actions recently added support for caching: https://github.com/actions/cache
This is a first step, caching composer dependencies.

Cannot be merged until https://github.com/actions/cache/issues/12 is resolved

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -56,6 +56,14 @@ jobs:
         run: echo 'memory_limit=-1' >> /usr/local/etc/php/php.ini
       - name: Install Composer
         run: wget -qO - https://raw.githubusercontent.com/composer/getcomposer.org/$COMPOSER_INSTALLER_COMMIT/web/installer | php -- --install-dir=/usr/local/bin --filename=composer --quiet
+      - name: Cache composer dependencies
```
Comment (teohhanhui) [inline .github/workflows/ci.yml]:
```suggestion
      - name: Cache Composer packages
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -56,6 +56,14 @@ jobs:
...
+        uses: actions/cache@preview
+        with:
+          path: ~/.composer/cache/files
+          key: composer-${{ matrix.php }}-${{ github.sha }}
```
Comment (teohhanhui) [inline .github/workflows/ci.yml]:
```suggestion
          key: composer-php${{ matrix.php }}-${{ github.sha }}
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -243,6 +259,14 @@ jobs:
         run: echo 'memory_limit=-1' >> /usr/local/etc/php/php.ini
       - name: Install Composer
         run: wget -qO - https://raw.githubusercontent.com/composer/getcomposer.org/$COMPOSER_INSTALLER_COMMIT/web/installer | php -- --install-dir=/usr/local/bin --filename=composer --quiet
+      - name: Cache composer dependencies
+        uses: actions/cache@preview
+        with:
+          path: ~/.composer/cache/files
+          key: composer-${{ matrix.php }}-prefer-lowest-${{ github.sha }}
```
Comment (teohhanhui) [inline .github/workflows/ci.yml]:
```suggestion
          key: composer-php${{ matrix.php }}-lowest-deps-${{ github.sha }}
```

**Comment by teohhanhui** [comment]
Blocked by https://github.com/actions/cache/issues/12

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -56,6 +56,14 @@ jobs:
         run: echo 'memory_limit=-1' >> /usr/local/etc/php/php.ini
       - name: Install Composer
         run: wget -qO - https://raw.githubusercontent.com/composer/getcomposer.org/$COMPOSER_INSTALLER_COMMIT/web/installer | php -- --install-dir=/usr/local/bin --filename=composer --quiet
+      - name: Cache Composer packages
+        uses: actions/cache@preview
+        with:
+          path: ~/.composer/cache/files
+          key: composer-php${{ matrix.php }}-${{ github.sha }}
```
Comment (localheinz) [inline .github/workflows/ci.yml]: You might want to take a look at https://github.com/localheinz/php-library-template/pull/131, perhaps this is better:

```suggestion
          key: composer-php${{ matrix.php }}-${{ hashFiles('**/composer.json') }}
```

Comment (teohhanhui) [reply .github/workflows/ci.yml]: That is wrong. It'll prevent the cache from being saved as long as composer.json doesn't change.

Comment (localheinz) [reply .github/workflows/ci.yml]: Works fine for me.

Comment (bendavies) [reply .github/workflows/ci.yml]: @localheinz are you sure? do you mean using `composer.lock` works fine for you?

**Comment by teohhanhui** [comment]
phpstan errors are unrelated.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -1147,6 +1251,15 @@ jobs:
         run: echo 'memory_limit=-1' >> /usr/local/etc/php/php.ini
       - name: Install Composer
         run: wget -qO - https://raw.githubusercontent.com/composer/getcomposer.org/$COMPOSER_INSTALLER_COMMIT/web/installer | php -- --install-dir=/usr/local/bin --filename=composer --quiet
+      - name: Cache Composer packages
+        uses: actions/cache@v1
+        with:
+          path: ~/.composer/cache/files
+          key: composer-php${{ matrix.php }}-${{ matrix.symfony }}-${{ github.sha }}
```
Comment (teohhanhui) [inline .github/workflows/ci.yml]:
```suggestion
          key: composer-php${{ matrix.php }}-symfony${{ matrix.symfony }}-${{ github.sha }}
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -1147,6 +1251,15 @@ jobs:
...
+          restore-keys: |
+            composer-php${{ matrix.php }}-${{ matrix.symfony }}-
```
Comment (teohhanhui) [inline .github/workflows/ci.yml]:
```suggestion
            composer-php${{ matrix.php }}-symfony${{ matrix.symfony }}-
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -243,6 +259,14 @@ jobs:
...
       - name: Install Composer
         run: wget -qO - https://raw.githubusercontent.com/composer/getcomposer.org/$COMPOSER_INSTALLER_COMMIT/web/installer | php -- --install-dir=/usr/local/bin --filename=composer --quiet
+      - name: Cache Composer packages
+        uses: actions/cache@v1
+        with:
+          path: ~/.composer/cache/files
+          key: composer-php${{ matrix.php }}-lowest-deps-${{ github.sha }}
+          restore-keys: |
+            composer-php${{ matrix.php }}-lowest-deps-
+            composer-
```
[the remark this answers is not in the document]
Comment (bendavies) [reply .github/workflows/ci.yml]: good spot

**Comment by teohhanhui** [comment]
Thanks @bendavies! 🎉 🚀

**Pull request merged**